src/units/abilities.cpp

```diff
@@ -1307,38 +1307,29 @@ unit_ability_list attack_type::overwrite_special_checking(const std::string& abi
 			overwrite_either = true;
 			break;
 		}
-		if(overwrite_special_affects(*i.ability_cfg) && (special_active(*i.ability_cfg, AFFECT_SELF, ability, "filter_student") || special_active_impl(other_attack_, shared_from_this(), *i.ability_cfg, AFFECT_OTHER, ability, "filter_student"))) {
+		if((*i.ability_cfg)["overwrite_specials"] == "one_side" && !overwrite_self) {
 			overwrite_self = true;
 		}
 	}
 	if(!overwrite_either && !overwrite_self){
 		return temp_list;
-	} else if(overwrite_either){
-		for(unit_ability_list::iterator i = temp_list.begin(); i != temp_list.end();) {
-			bool overwrite = false;
-			if(filter_self == "filter_student"){
-				overwrite = overwrite_special_affects(*i->ability_cfg);
-			}
-			if(!overwrite && (special_active(*i->ability_cfg, AFFECT_SELF, ability, filter_self) || special_active_impl(other_attack_, shared_from_this(), *i->ability_cfg, AFFECT_OTHER, ability, filter_self))) {
-				i = temp_list.erase(i);
-			} else {
-				++i;
-			}
+	}
+
+	// At this point we need to return a changed list, so create a non-const one to return
+	unit_ability_list overwrite_list;
+	for(const auto& i : temp_list) {
+		bool overwrite = false;
+		if(!is_special && overwrite_either){
+			overwrite = overwrite_special_affects(*i.ability_cfg);
+		} else if(overwrite_self){
+			bool temp_is_ability = !is_special ? overwrite_special_affects(*i.ability_cfg) : false;
```

Comment:
I think that ternary simplifies to just `(!is_special) && overwrite_specials_affects(*i.ability_cfg)`.

Reply:
done here also